### ISSUE TITLE
Database changes - BoltDB is archived; allow db.Save(); return ID on hooks

### ISSUE THREAD

**Issue by nanohard** [ponzu-cms/ponzu#270]
[BoltDB](https://github.com/boltdb/bolt) is now archived and unmaintained.
The author suggests looking at [bbolt](https://github.com/coreos/bbolt).
It seems to be a drop-in replacement:
> bbolt includes bug fixes, performance enhancements, and features not found in Bolt while preserving backwards compatibility with the Bolt API.

Now, onto moar.
[Storm](https://github.com/asdine/storm) is a wrapper for [bbolt](https://github.com/coreos/bbolt) (they switched from BoltDB when it got archived), which would allow for db.Save() and some other neat wrappers, which is something that is sorely needed. I had to implement some hackery to get around not being able to update a db entry on a hook; we need to allow for this.

This leads me to hooks in Ponzu; for all `afterXyzCreate()` interfaces we should be returning the ID of the entry.

Please leave your thoughts and candy... or just candy.

**Comment by nilslice** [comment]
🍬 

💭 
> BoltDB is now archived and unmaintained. 

True. Event though it has been this way for quite some time, the bolt API and codebase remains stable as far as I can tell. So, I'm not worried yet about swapping these dependencies, but would like to further investigate how `bbolt` could bring additional value to Ponzu. Glad you have looked into this, and I appreciate the call-out. 

There should definitely be more emphasis put on the `db` package for Ponzu, and adding some tests around the boltdb usage is probably something that should be done soon.

> [...] This leads me to hooks in Ponzu; for all afterXyzCreate() interfaces we should be returning the ID of the entry.

That's a great idea. I think it would be a useful addition to these hooks: 
```go
AfterAPICreate(http.ResponseWriter, *http.Request) error
AfterAdminCreate(http.ResponseWriter, *http.Request) error
AfterApprove(http.ResponseWriter, *http.Request) error
AfterEnable(http.ResponseWriter, *http.Request) error // I'll have to think about this one a bit more, since the addon API is likely to change a bit
```

Any other hooks I miss? Let me know if you think so.

Re: moar.
I'd like to avoid introducing the dependency on Storm. Not that it isn't a useful tool, but keeping the core database code as low-level and close to the stdlib as possible is a goal. I'm sure there would be things that could be done easier using a wrapper for bolt's API, but I think it is still important to have as few dependencies between the Go stdlib and Ponzu's database layer. I could be convinced otherwise though, as I have not tried Storm or other wrappers for Ponzu. If you feel compelled to, feel free to try it out though and share your code/thoughts.

**Comment by nanohard** [comment]
**Hooks** `system/item/item.go`
These hooks should be returning the ID, because the ID is available. I can think of several situations where you'd want to manipulate something based on what gets returned. If the user wants the whole object, they would then have the ID to query. I am already returning the ID for `AfterAdminCreate` in my customized Ponzu:
```go
AfterAPICreate()

AfterAdminCreate()

BeforeAPIUpdate()
AfterAPIUpdate()

BeforeAdminUpdate()
AfterAdminUpdate()

BeforeSave()  // if applicable
AfterSave()

BeforeApprove()  // if applicable
AfterApprove()

BeforeReject()  // if applicable
AfterReject()  // if applicable

BeforeDelete()
```

**Storm**
My thought process for switching to Storm.
TL;DR - less work, lower barrier of entry:

- Uses `bbolt`, which is a backward-compatible API for `BoltDB` with fixes and upgrades.
- Can still use `Bolt` directly if needed.
- Why force each user to create their own wrapper functions? It's not like anyone will _not_ need to save or query the database on their own terms.
- Storm has a built-in codec selector for marshalling and unmarshalling, making it possible to easily use `gob`, which is a golang-to-golang encoding stream; it's like JSON, but magnitudes faster for communicating with other golang tools or systems. Could come in handy for high-concurrency communications.
- If Ponzu takes the course of allowing to use other databases besides Bolt, then the wrappers for Bolt functions are already there, making it easier.
- Storm uses 16 packages:
  - 5 are from its own internal packages
  - 1 is a third-party (bbolt, the database package)
  - all others are from the stdlib

I don't mind creating an alternate version of Ponzu that implements Storm to try out and report on. I would keep it private so as to not cause any confusion. But this is a big alteration and would take some time, especially the testing/timing comparison as I'm unfamiliar with test driven development (this would force me to use tests).

**Comment by nilslice** [comment]
I'll take a closer look at Storm, but re: the hooks -- I believe that all, aside for the ones I mentioned, have the ID available on the pointer receiver already. Being lifecycle hooks, they are called on objects mid-request, whom have already been hydrated by data from the DB. 

I could be wrong about some, but we could easily double-check.. Unless I'm being a jerk and you have already performed these tests, and determined that they do not have the ID available 😆 

For example, in the case of the following:
```go
type MyContent struct {
	item.Item

	Name string `json:"string"`
	// ...
}

func (m *MyContent) BeforeAPIUpdate(http.ResponseWriter, *http.Request) error {
	fmt.Println(m.ID) // would print the actual ID, since it was set
}
```

The setting of the fields is done above the assertion to the `item.Hookable` interface and use of the hook method here: 
https://github.com/ponzu-cms/ponzu/blob/e43cda203b2b3eadab6bea1891e83fe232555fb5/system/api/update.go#L61-L171

**Comment by gedw99** [comment]
what about https://github.com/timshannon/bolthold ?

Its wraps bbolt but still gives full access to bolt.

One thing we really need is a way for Bolt to be clustered for HA purposes.
There are a few Bolt Raft implementations out there.

**Comment by olliephillips** [comment]
Going to close this issue. Reading through, my own thoughts are that BoltDB is sufficient for the Ponzu CMS project at this time. Please re-open if needs be.
